resolwe/flow/managers/base.py

```diff
@@ -624,39 +628,47 @@ def _communicate(self, run_sync=False, verbosity=1, executor='resolwe.flow.execu
                         data.status = Data.STATUS_WAITING
                     data.save(render_name=True)
 
-                    if program is not None:
-                        try:
-                            priority = 'normal'
-                            if data.process.persistence == Process.PERSISTENCE_TEMP:
-                                # TODO: This should probably be removed.
-                                priority = 'high'
-                            if data.process.scheduling_class == Process.SCHEDULING_CLASS_INTERACTIVE:
-                                priority = 'high'
-
-                            program = self._include_environment_variables(program)
-
-                            data_dir = self._prepare_data_dir(data.id)
-                            executor_module, runtime_dir = self._prepare_executor(data.id, executor)
-                            self._prepare_context(data.id, data_dir, runtime_dir, verbosity=verbosity)
-                            self._prepare_script(runtime_dir, program)
-                            argv = [
-                                '/bin/bash',
-                                '-c',
-                                self.settings_actual.get('FLOW_EXECUTOR', {}).get('PYTHON', '/usr/bin/env python') +
-                                ' -m executors ' + executor_module
-                            ]
-                        except OSError as err:
-                            logger.error(__(
-                                "OSError occurred while preparing data {} (will skip): {}",
-                                data.id, err
-                            ))
-                            continue
-
-                        queue.append((data.id, priority, runtime_dir, argv))
+                    # Signal transaction success through the flag.
+                    transaction.on_commit(lambda: transaction_success.append(True))
```

Comment:
This is quite convoluted, why don't you instead push to queue on transaction success instead of introducing an intermediate flag-list?

Reply:
This was an explicit scoping decision. When an object is vetted for execution, there's still quite a bit of code to execute, but with enough state in the local scope to make a separate function awkward. The flag list is necessary because lambdas can't contain assignment expressions, so a primitive wouldn't work. Given your comment though, we could refactor this entirely so that `_communicate` only does the scanning and then calls a special commit function `on_commit`. This way, it would be easier to correctly handle transactions and get rid of the queue. Thoughts?

Reply:
> Given your comment though, we could refactor this entirely so that _communicate only does the scanning and then calls a special commit function on_commit. This way, it would be easier to correctly handle transactions and get rid of the queue. Thoughts?

This sounds better. So the run methods would be called one by one (after commit) instead of first going to the queue?